tests/unit/s2n_cipher_suite_match_test.c

```diff
@@ -850,5 +850,543 @@ int main(int argc, char **argv)
+                    .allow_chacha20_boosting = true,
+                };
+
+                uint8_t wire[] = {
+                    /* Client signalled chacha20 boosting; not negotiated as server does not have any chacha20 ciphersuites */
+                    TLS_CHACHA20_POLY1305_SHA256,
+                    /* Not negotiated as the server prefers aes 128 over aes 256 */
+                    TLS_AES_256_GCM_SHA384,
+                    /* Negotiated */
+                    TLS_AES_128_GCM_SHA256,
+                };
+                uint8_t count = sizeof(wire) / S2N_TLS_CIPHER_SUITE_LEN;
+
+                /* Verify that server is able to negotiate aes_128 even without any chacha20 ciphersuites in its preferences */
+                cipher_preferences.allow_chacha20_boosting = false;
+                EXPECT_SUCCESS(s2n_set_cipher_as_tls_server(connection, wire, count));
+                EXPECT_EQUAL(connection->secure->cipher_suite, &s2n_tls13_aes_128_gcm_sha256);
+                EXPECT_SUCCESS(s2n_disable_tls13_in_test());
+            }
+
+            /* Test chacha20 boosting when the most preferred ciphersuite fails version validation */
+            {
+                DEFER_CLEANUP(struct s2n_connection *connection = s2n_connection_new(S2N_SERVER), s2n_connection_ptr_free);
+                EXPECT_NOT_NULL(connection);
+
+                connection->security_policy_override = &security_policy;
+                connection->kex_params.server_ecc_evp_params.negotiated_curve = s2n_all_supported_curves_list[0];
+                EXPECT_SUCCESS(s2n_connection_set_all_protocol_versions(connection, S2N_TLS12));
+
+                EXPECT_SUCCESS(s2n_connection_set_config(connection, server_config));
+
+                static struct s2n_cipher_suite *test_cipher_suite_list[] = {
+                    /* Only negotiated if chacha20 boosting is off */
+                    &s2n_ecdhe_ecdsa_with_aes_128_cbc_sha256,       
+                    /* Invalid; never use tls 1.3 ciphers on pre-tls 1.3 connections */
+                    &s2n_tls13_chacha20_poly1305_sha256,
+                    /* Negotiated (if chacha20 boosting is on) */
+                    &s2n_ecdhe_rsa_with_chacha20_poly1305_sha256,   
+                };
+
+                cipher_preferences = (struct s2n_cipher_preferences) {
+                    .count = s2n_array_len(test_cipher_suite_list),
+                    .suites = test_cipher_suite_list,
+                    .allow_chacha20_boosting = true,
+                };
+
+                uint8_t wire[] = {
+                    /* Client signalled chacha20 boosting. Not negotiated by server since it is invalid for the connection. */
+                    TLS_CHACHA20_POLY1305_SHA256,
+                    /* Negotiated if chacha20 boosting is off */
+                    TLS_ECDHE_ECDSA_WITH_AES_128_CBC_SHA256,
+                    /* Negotiated if chacha20 boosting is off */
+                    TLS_ECDHE_RSA_WITH_CHACHA20_POLY1305_SHA256
```

Comment:
I'm guessing this is a typo? We negotiate TLS_ECDHE_ECDSA_WITH_AES_128_CBC_SHA256 if chacha20 boosting is off, TLS_ECDHE_RSA_WITH_CHACHA20_POLY1305_SHA256 if chacha20 boosting is on.

Reply:
Yeah you're right. Fixed typo comment.